File: app/src/main/java/net/theluckycoder/advancedreboot/MainActivity.java
package net.theluckycoder.advancedreboot;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.app.AppCompatDelegate;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import com.google.firebase.analytics.FirebaseAnalytics;

import eu.chainfire.libsuperuser.Shell;

public final class MainActivity extends AppCompatActivity {

    static { AppCompatDelegate.setCompatVectorFromResourcesEnabled(true); }

    private FirebaseAnalytics mFirebaseAnalytics;
    private InterstitialAd mInterstitialAd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setTitle(R.string.app_name);

        if (Build.MANUFACTURER.matches("samsung"))
            findViewById(R.id.bootloaderReboot).setVisibility(View.GONE);
        else
            findViewById(R.id.downloadReboot).setVisibility(View.GONE);

        checkForRoot();

        //Init AdMob
        AdView mAdView = findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice("304C7D4CF3DD1D2C556771826CCF9037")
                .build();
        mAdView.loadAd(adRequest);

        mInterstitialAd = new InterstitialAd(this);
        mInterstitialAd.setAdUnitId("ca-app-pub-1279472163660969/3259304920");
        mInterstitialAd.setAdListener(new AdListener() {
            @Override
            public void onAdClosed() {
                requestNewInterstitial();
            }
        });
        requestNewInterstitial();

        // Obtain the FirebaseAnalytics instance.
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    public void reboot(View view) {
        int id = view.getId();

        Bundle params = new Bundle();

        switch (id) {
            case R.id.normalReboot:
                confirmDialog(new String[]{"reboot"});
                params.putString("type", "normal");
                break;
            case R.id.fastReboot:
                confirmDialog(new String[]{"setprop ctl.restart zygote"});
                params.putString("type", "fast");
                break;
            case R.id.recoveryReboot:
                confirmDialog(new String[]{"reboot recovery"});
                params.putString("type", "recovery");
                break;
            case R.id.bootloaderReboot:
                confirmDialog(new String[]{"reboot bootloader"});
                params.putString("type", "bootloader");
                break;
            case R.id.downloadReboot:
                confirmDialog(new String[]{"reboot download"});
                params.putString("type", "bootloader");
                break;
            case R.id.safeModeReboot:
                confirmDialog(new String[]{"setprop persist.sys.safemode 1", "reboot"});
                params.putString("type", "safeMode");
                break;
            case R.id.interfaceReboot:
                confirmDialog(new String[]{"pkill com.android.systemui"});
                params.putString("type", "interface");
                break;
            case R.id.shutdownReboot:
                confirmDialog(new String[]{"reboot -p"});
                params.putString("type", "shutdown");
                break;
        }

        mFirebaseAnalytics.logEvent("reboot", params);
    }

    private void checkForRoot() {
        if (!Shell.SU.available()) {
            AlertDialog.Builder dialog = new AlertDialog.Builder(this);
            dialog.setTitle(R.string.root_required);
            dialog.setMessage(R.string.root_required_desc);
            dialog.setCancelable(false);
            dialog.setNegativeButton(R.string.quit, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    finish();
                }
            });
            dialog.setPositiveButton(R.string.restart_app, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    recreate();
                }
            });
            dialog.show();
        }
    }

    private void confirmDialog(final String[] commands) {
        if (PreferenceManager.getDefaultSharedPreferences(this).getBoolean("confirm_reboots", true)) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.confirm_continue);
            builder.setMessage(R.string.confirm_continue_desc);
            builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    showProgressDialog(commands);
                }
            });
            builder.setNegativeButton(android.R.string.no, null);
            builder.show();

            if(mInterstitialAd.isLoaded())
                mInterstitialAd.show();
        } else
            showProgressDialog(commands);
    }

    private void showProgressDialog(String[] commands) {
        new RebootTask(commands).execute();
        ProgressDialog progressDialog = new ProgressDialog(this);
        progressDialog.setTitle(getString(R.string.rebooting_device));
        progressDialog.setCancelable(false);
        progressDialog.setIndeterminate(true);
        progressDialog.show();
    }

    private void requestNewInterstitial() {
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice("304C7D4CF3DD1D2C556771826CCF9037")
                .build();

        mInterstitialAd.loadAd(adRequest);
    }

}

File: app/src/main/java/net/theluckycoder/advancedreboot/MainActivity.kt
package net.theluckycoder.advancedreboot

import android.app.ProgressDialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.preference.PreferenceManager
import eu.chainfire.libsuperuser.Shell
import java.util.Locale

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        if (PreferenceManager.getDefaultSharedPreferences(this).getBoolean("dark_theme", false)) {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        } else {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        }

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (Build.MANUFACTURER.toString().toLowerCase(Locale.ROOT) == "samsung") {
            findViewById<View>(R.id.bootloaderReboot).visibility = View.GONE
        } else {
            findViewById<View>(R.id.downloadReboot).visibility = View.GONE
        }

        checkForRoot()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == 8) {
            val useDarkTheme = PreferenceManager.getDefaultSharedPreferences(this).getBoolean("dark_theme", false)

            if (!useDarkTheme && AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_NO) {
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            } else if (useDarkTheme && AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_YES) {
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            }

            recreate()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_settings) {
            startActivityForResult(Intent(this, SettingsActivity::class.java), 8)
        }
        return super.onOptionsItemSelected(item)
    }

    fun reboot(view: View) {
        when (view.id) {
            R.id.normalReboot -> confirmDialog(listOf("reboot"))
            R.id.fastReboot -> confirmDialog(listOf("setprop ctl.restart zygote"))
            R.id.recoveryReboot -> confirmDialog(listOf("reboot recovery"))
            R.id.bootloaderReboot -> confirmDialog(listOf("reboot bootloader"))
            R.id.downloadReboot -> confirmDialog(listOf("reboot download"))
            R.id.safeModeReboot -> confirmDialog(listOf("setprop persist.sys.safemode 1", "reboot"))
            R.id.interfaceReboot -> confirmDialog(listOf("pkill com.android.systemui"))
            R.id.shutdownReboot -> confirmDialog(listOf("reboot -p"))
        }
    }

    private fun checkForRoot() {
        if (BuildConfig.DEBUG || Shell.SU.available()) return

        AlertDialog.Builder(this)
            .setTitle(R.string.root_required)
            .setMessage(R.string.root_required_desc)
            .setCancelable(false)
            .setNegativeButton(R.string.quit) { _, _ -> finish() }
            .setPositiveButton(R.string.restart_app) { _, _ -> recreate() }
            .show()
    }

    private fun confirmDialog(commands: List<String>) {
        if (PreferenceManager.getDefaultSharedPreferences(this).getBoolean("confirm_reboots", true)) {
            AlertDialog.Builder(this)
                .setTitle(R.string.confirm_continue)
                .setMessage(R.string.confirm_continue_desc)
                .setPositiveButton(android.R.string.yes) { _, _ -> showProgressDialog(commands) }
                .setNegativeButton(android.R.string.no, null)
                .show()
        } else {
            showProgressDialog(commands)
        }
    }

    private fun showProgressDialog(commands: List<String>) {
        RebootTask.execute(commands)

        @Suppress("DEPRECATION")
        ProgressDialog(this).apply {
            setTitle(getString(R.string.rebooting_device))
            setCancelable(false)
            isIndeterminate = true
            show()
        }
    }
}
